engine/engine.go

```diff
@@ -0,0 +1,217 @@
+package engine
+
+import (
+	"errors"
+	"strings"
+	"time"
+
+	"github.com/RTradeLtd/Lens/models"
+	"github.com/go-ego/riot"
+	"github.com/go-ego/riot/types"
+	"go.uber.org/zap"
+)
+
+// Searcher exposes Engine's primary functions
+type Searcher interface {
+	Index(doc Document)
+	Search(query Query) ([]Result, error)
+
+	IsIndexed(hash string) bool
+	Remove(hash string)
+}
+
+// Engine implements Lens V2's core search functionality
+type Engine struct {
+	e *riot.Engine
+	l *zap.SugaredLogger
+
+	stop chan bool
+}
+
+// Opts denotes options for the Lens engine
+type Opts struct {
+	DictPath  string
+	StorePath string
+}
+
+// New instantiates a new Engine
+func New(l *zap.SugaredLogger, opts Opts) (*Engine, error) {
+	var e = &riot.Engine{}
+	var r = types.EngineOpts{}
+
+	// database persistence settings
+	if opts.StorePath != "" {
+		r.UseStore = true
+		r.StoreEngine = "bg" // use badger for persistence
+		r.StoreFolder = opts.StorePath
+	}
+
+	l.Infow("starting up search engine", "config", r)
+	e.Init(r)
+	return &Engine{
+		e: e,
+		l: l,
+	}, nil
+}
+
+// Run initiates period index flushes
+func (e *Engine) Run(indexInterval time.Duration) {
+	var ticker = time.NewTicker(indexInterval)
+	e.stop = make(chan bool)
+	for {
+		select {
+		case <-ticker.C:
+			var now = time.Now()
+			e.e.Flush()
```

Comment:
Does flushing write to disk?

Reply:
According to the documentation, this only rebuilds the index. Doing it periodically seemed like the right way to do it, but I'm not sure yet - perhaps better to flush after actual write operations. Will keep reading into this

Reply:
Yes that seems like it would be a good idea. We may want to look at some method of batching together flushes to save against disk IO

Reply:
The implmentation here batches flushes together, but I don't think it's a significant penalty to do it per index